controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -184,14 +184,14 @@ private boolean changesAvailable(Application application, JobStatus previous, Jo
         if ( ! application.deploying().isPresent()) return false;
         Change change = application.deploying().get();
 
-        if ( ! previous.lastSuccess().isPresent() && 
+        if ( ! previous.isSuccess() &&
```

Comment:
I believe this must have been the intention?

Reply:
Actually, no. job.isSuccess tells us if the job is currently successful. This checks if there is a successful outcome available, regardless of whether it is the current state.

Reply:
The combination ` ! lastSuccess().isPresent() && ! production...` could be reduced to just ` ! lastSuccess().isPresent()`, then, because this check is also done in the latter.

Reply:
So something's not right ...

Reply:
Also, reading the doc comment for the `production...` check leads me to believe this was, indeed, the intention:

_If the job is failing, and isn't a production job with a previous success on this version, bail out._

Reply:
Where is the code you are talking about now?

Reply:
On the line below.

Reply:
Below what?

Reply:
Lines 187 (marked, above) and 188 have a compound condition:
```
if ( ! previous.lastSuccess().isPresent() &&
     ! productionUpgradeHasSucceededFor(...))
```

Reply:
And you think that condition could be reduced to just the first line, is that it?

Reply:
As it stands, yes:

* If `last success` is present, the first condition becomes `! true`, and we short-circuit.  
* If `last success` is not present, we check whether `! productionUpgradeHasSuccededFor`; this is also `true`, since it, too, checks if `last success` is present.

This, and the javadoc for `productionUpgradeHasSuccededFor`, as well as general reasoning, lead to my conclusion about the intention here.

Reply:
I think you mean we can drop the *first* condition, not the last then?

Reply:
No, that would make us always bail out for non-production jobs and non-version-changes. 

I didn't say we should drop anything, I said as it stands, we could as well omit the last condition, which leads me to believe the current condition is faulty.

Reply:
Perhaps there would have been no confusion if my highlight covered the line with the second condition as well...

Reply:
... since it is left untouched.

Reply:
Ok, now I get it.
Yes, I think we can remove the second condition as things stand now.

Reply:
Agreed.